vazco/uniforms
# Replace `lerna` with `turborepo`

### Description

This is minimal setup of `turborepo`. It replaces `lerna` package. I focused on keeping it as simple as possible, so current configuration can be (and should be) updated in the future. For example, `eslint` and `typescript` configs can be moved to internal packages ([docs](https://turbo.build/repo/docs/guides/tools/eslint)). 

During implementation, I had issues with `reproductions` and `website` projects and couldn't make them work. They have been removed for now in order to create this *draft* pull request. Removed files in the `reproductions`, `website` and `docs` can be ignored in the code review process.

### Build process

The biggest change is a new building process. To take full advantage of the new project structure, I changed the build process from incrementally building with `tsc` in the root directory to running each packages' `tsc` build script in parallel. It's faster on cold start, but also incredibly fast when it comes to cache utilization. There are results after running it with Node `v22.8.0` and npm `v10.8.3` (cold start):

| | master | turborepo |
| - | :-: | :-: |
| `build` | 49s | 21s |
| `test` | 37s | 14s |

### Missing features (to be implemented)

- [X] `reproductions` project
- [ ] Documentation (`website` & `docs` directories)
- [ ] Publishing process

`lerna` had already implemented publishing process (package versioning and publishing them to the NPM registry). `turborepo` doesn't have such feature. It is recommended to use [`changesets`](https://github.com/changesets/changesets) ([docs](https://turbo.build/repo/docs/guides/publishing-libraries#versioning-and-publishing)).

I suggest to implement docs and publishing process in other PRs.

### List of changes

- Removed `lerna` package
- Added `turbo` package
- Changed build process
- Removed reproductions and docs (temporary)
- Added `devDependencies` and moved some packages from `dependencies`. It can be refactored further though.
- Simplified `tsconfig` files to remove incremental build
- Added `simple-schema.d.ts` to the `uniforms-bridge-simple-schema-2`. This file was originally located in the `website` directory. This directory was temporary removed and missing file was causing type issues during testing process.
- Removed `publish` script as it was based on `lerna`
- Removed unused packages in the root directory

### How to test changes

Use Node `v22` (check current version in the `.nvmrc`) and npm `v10` (check current version in the `package.json -> packageManager`)

1. Checkout to `turborepo` branch
2. Run `npm run clean` (it deletes packages' builds)
3. Run `npm run reset` (it deletes `node_modules`, turbo cache, etc)
4. Run `npm ci` (it also runs `build` script)
5. To run manually building process for packages - run `npm run build`
6. Run `npm test`
